File: src/e2e/restream_test.go
//go:build e2e

package e2e

import (
	"fmt"
	"net/http"
	"path/filepath"
	"testing"
	"time"

	"Theatrum/constants"
)

func TestRestream_HLSPassthrough(t *testing.T) {
	testVideoPath := testVideoPath(t)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/restream/mystream":
    stream:
      type: restream
      source_url: "%s"
      path: "restream/mystream"
      distribution:
        hls:
          segment_duration: 2
          window_size: 5
`, testVideoPath),
	})

	// Wait for the playlist to appear on disk
	// Passthrough streams go into {path}/default/
	playlistPath := filepath.Join(ts.DataDir, "restream", "mystream", constants.DefaultQuality, constants.SubPlaylist)
	waitForFile(t, playlistPath, 20*time.Second)

	// Wait for at least one segment
	segmentPath := filepath.Join(ts.DataDir, "restream", "mystream", constants.DefaultQuality, "segment_000.ts")
	waitForFile(t, segmentPath, 20*time.Second)

	// Fetch the master playlist via HTTP
	masterURL := fmt.Sprintf("http://%s/restream/mystream/%s", ts.HTTPAddr, constants.MasterPlaylist)
	resp := waitForHTTP(t, masterURL, 15*time.Second)
	resp.Body.Close()

	status, body := httpGetBody(t, masterURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for master playlist, got %d", status)
	}
	m3u8 := parseM3U8(body)
	if !m3u8.IsMaster {
		t.Error("Expected master playlist")
	}

	// Fetch the sub playlist via HTTP
	subPlaylistURL := fmt.Sprintf("http://%s/restream/mystream/%s/%s", ts.HTTPAddr, constants.DefaultQuality, constants.SubPlaylist)
	status, body = httpGetBody(t, subPlaylistURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for sub playlist, got %d", status)
	}
	m3u8 = parseM3U8(body)
	if m3u8.SegmentCount == 0 {
		t.Error("Expected at least one segment in playlist")
	}

	// Verify live cache headers (restream is treated as live for caching)
	resp = httpGet(t, subPlaylistURL)
	defer resp.Body.Close()
	cacheControl := resp.Header.Get("Cache-Control")
	if cacheControl != "no-cache, no-store, must-revalidate" {
		t.Errorf("Expected no-cache for restream playlist, got %q", cacheControl)
	}
}

func TestRestream_HLSMultiQuality(t *testing.T) {
	testVideoPath := testVideoPath(t)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/restream/premium":
    stream:
      type: restream
      source_url: "%s"
      path: "restream/premium"
      qualities:
        low:
          width: 640
          height: 360
          framerate: 24
          bitrate: "800k"
          codec: "libx264"
          audio:
            bitrate: "96k"
            codec: "aac"
      distribution:
        hls:
          segment_duration: 2
          window_size: 5
`, testVideoPath),
	})

	// Wait for master.m3u8 to appear (multi-quality creates it via FFmpeg)
	masterPath := filepath.Join(ts.DataDir, "restream", "premium", constants.MasterPlaylist)
	waitForFile(t, masterPath, 30*time.Second)

	// Wait for at least one quality subdir to appear
	lowPlaylistPath := filepath.Join(ts.DataDir, "restream", "premium", "low", constants.SubPlaylist)
	waitForFile(t, lowPlaylistPath, 30*time.Second)

	// Fetch master playlist via HTTP
	masterURL := fmt.Sprintf("http://%s/restream/premium/%s", ts.HTTPAddr, constants.MasterPlaylist)
	status, body := httpGetBody(t, masterURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for master playlist, got %d", status)
	}
	m3u8 := parseM3U8(body)
	if !m3u8.IsMaster {
		t.Error("Expected master playlist")
	}
	if len(m3u8.StreamPlaylists) == 0 {
		t.Error("Expected at least one stream playlist in master")
	}
}

func TestRestream_DASHPassthrough(t *testing.T) {
	testVideoPath := testVideoPath(t)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/restream/dash":
    stream:
      type: restream
      source_url: "%s"
      path: "restream/dash"
      distribution:
        dash:
          segment_duration: 2
          window_size: 5
`, testVideoPath),
	})

	// Wait for manifest.mpd to appear on disk
	manifestPath := filepath.Join(ts.DataDir, "restream", "dash", constants.DashManifest)
	waitForFile(t, manifestPath, 20*time.Second)

	// Fetch manifest via HTTP
	manifestURL := fmt.Sprintf("http://%s/restream/dash/%s", ts.HTTPAddr, constants.DashManifest)
	resp := waitForHTTP(t, manifestURL, 15*time.Second)
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		t.Fatalf("Expected 200 for manifest, got %d", resp.StatusCode)
	}

	contentType := resp.Header.Get("Content-Type")
	if contentType != "application/dash+xml" {
		t.Errorf("Expected Content-Type application/dash+xml, got %q", contentType)
	}
}

func TestRestream_ViewerTracking(t *testing.T) {
	testVideoPath := testVideoPath(t)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/restream/tracked":
    stream:
      type: restream
      source_url: "%s"
      path: "restream/tracked"
      distribution:
        hls:
          segment_duration: 2
          window_size: 5
      viewers:
        enabled: true
        window: 1
      views:
        enabled: true
        window: 1
`, testVideoPath),
	})

	// Wait for the playlist to appear
	playlistPath := filepath.Join(ts.DataDir, "restream", "tracked", constants.DefaultQuality, constants.SubPlaylist)
	waitForFile(t, playlistPath, 20*time.Second)

	// Check that viewers.txt endpoint works
	viewersURL := fmt.Sprintf("http://%s/restream/tracked/%s", ts.HTTPAddr, constants.ViewersFile)
	status, body := httpGetBody(t, viewersURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for viewers.txt, got %d", status)
	}
	if body != "0" {
		t.Errorf("Expected 0 viewers initially, got %q", body)
	}

	// Check that views.txt endpoint works
	viewsURL := fmt.Sprintf("http://%s/restream/tracked/%s", ts.HTTPAddr, constants.ViewsFile)
	status, body = httpGetBody(t, viewsURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for views.txt, got %d", status)
	}
	if body != "0" {
		t.Errorf("Expected 0 views initially, got %q", body)
	}
}

File: src/test/java/com/fjourdren/theatrum/e2e/RestreamTest.java
package com.fjourdren.theatrum.e2e;

import com.fjourdren.theatrum.domain.constant.VideoConstants;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Port of Go's {@code restream_test.go}: channels that pull from an external source rather than
 * being pushed to. The source here is the test video on disk — any FFmpeg-readable URL works.
 * All four channels auto-start on boot, so one context covers every case.
 */
class RestreamTest {

    private static final String CHANNELS = """
              "/restream/mystream":
                stream:
                  type: restream
                  source_url: "%1$s"
                  path: "restream/mystream"
                  distribution:
                    hls:
                      segment_duration: 2
                      window_size: 5
              "/restream/premium":
                stream:
                  type: restream
                  source_url: "%1$s"
                  path: "restream/premium"
                  qualities:
                    low:
                      width: 640
                      height: 360
                      framerate: 24
                      bitrate: "800k"
                      codec: "libx264"
                      audio:
                        bitrate: "96k"
                        codec: "aac"
                  distribution:
                    hls:
                      segment_duration: 2
                      window_size: 5
              "/restream/dash":
                stream:
                  type: restream
                  source_url: "%1$s"
                  path: "restream/dash"
                  distribution:
                    dash:
                      segment_duration: 2
                      window_size: 5
              "/restream/tracked":
                stream:
                  type: restream
                  source_url: "%1$s"
                  path: "restream/tracked"
                  distribution:
                    hls:
                      segment_duration: 2
                      window_size: 5
                  viewers:
                    enabled: true
                    window: 1
                  views:
                    enabled: true
                    window: 1
            """;

    @TempDir
    static Path tempDir;

    static TestServerSupport server;

    @BeforeAll
    static void startServer() {
        TestServerSupport.requireFfmpeg();
        server = TestServerSupport.start(tempDir, CHANNELS.formatted(TestServerSupport.testVideo()));
    }

    @AfterAll
    static void stopServer() {
        if (server != null) {
            server.close();
        }
    }

    private Path restreamDir(String name) {
        return server.dataDir.resolve("restream").resolve(name);
    }

    @Test
    void hlsPassthroughServesAMasterAndSubPlaylist() {
        Path defaultDir = restreamDir("mystream").resolve(VideoConstants.DEFAULT_QUALITY);
        TestServerSupport.waitForFile(defaultDir.resolve(VideoConstants.SUB_PLAYLIST), Duration.ofSeconds(30));
        TestServerSupport.waitForFile(defaultDir.resolve("segment_000.ts"), Duration.ofSeconds(30));

        var master = server.waitForHttp("/restream/mystream/" + VideoConstants.MASTER_PLAYLIST,
                Duration.ofSeconds(20));
        assertThat(master.status()).isEqualTo(200);
        assertThat(TestServerSupport.parseM3u8(master.body()).master()).isTrue();

        String subUrl = "/restream/mystream/%s/%s".formatted(VideoConstants.DEFAULT_QUALITY,
                VideoConstants.SUB_PLAYLIST);
        var sub = server.get(subUrl);
        assertThat(sub.status()).isEqualTo(200);
        assertThat(TestServerSupport.parseM3u8(sub.body()).segmentCount()).isPositive();
        // A restream is treated as live for caching purposes.
        assertThat(sub.header("Cache-Control")).isEqualTo("no-cache, no-store, must-revalidate");
    }

    @Test
    void hlsMultiQualityServesAMasterPlaylistWithVariants() {
        TestServerSupport.waitForFile(restreamDir("premium").resolve(VideoConstants.MASTER_PLAYLIST),
                Duration.ofSeconds(40));
        TestServerSupport.waitForFile(restreamDir("premium").resolve("low").resolve(VideoConstants.SUB_PLAYLIST),
                Duration.ofSeconds(40));

        var master = server.waitForHttp("/restream/premium/" + VideoConstants.MASTER_PLAYLIST,
                Duration.ofSeconds(20));
        assertThat(master.status()).isEqualTo(200);

        var playlist = TestServerSupport.parseM3u8(master.body());
        assertThat(playlist.master()).isTrue();
        assertThat(playlist.streamPlaylists()).isNotEmpty();
    }

    @Test
    void dashPassthroughServesAManifest() {
        TestServerSupport.waitForFile(restreamDir("dash").resolve(VideoConstants.DASH_MANIFEST),
                Duration.ofSeconds(30));

        var response = server.waitForHttp("/restream/dash/" + VideoConstants.DASH_MANIFEST,
                Duration.ofSeconds(20));
        assertThat(response.status()).isEqualTo(200);
        assertThat(response.header("Content-Type")).isEqualTo("application/dash+xml");
    }

    @Test
    void exposesViewerAndViewCountersStartingAtZero() {
        TestServerSupport.waitForFile(
                restreamDir("tracked").resolve(VideoConstants.DEFAULT_QUALITY).resolve(VideoConstants.SUB_PLAYLIST),
                Duration.ofSeconds(30));

        var viewers = server.get("/restream/tracked/" + VideoConstants.VIEWERS_FILE);
        assertThat(viewers.status()).isEqualTo(200);
        assertThat(viewers.body()).isEqualTo("0");

        var views = server.get("/restream/tracked/" + VideoConstants.VIEWS_FILE);
        assertThat(views.status()).isEqualTo(200);
        assertThat(views.body()).isEqualTo("0");
    }
}
